Firestore/core/src/firebase/firestore/remote/datastore.mm

```diff
@@ -99,16 +120,102 @@
 
 std::shared_ptr<WatchStream> Datastore::CreateWatchStream(
     id<FSTWatchStreamDelegate> delegate) {
-  return std::make_shared<WatchStream>(worker_queue_, credentials_, serializer_,
+  return std::make_shared<WatchStream>(worker_queue_, credentials_,
+                                       serializer_bridge_.GetSerializer(),
                                        &grpc_connection_, delegate);
 }
 
 std::shared_ptr<WriteStream> Datastore::CreateWriteStream(
     id<FSTWriteStreamDelegate> delegate) {
-  return std::make_shared<WriteStream>(worker_queue_, credentials_, serializer_,
+  return std::make_shared<WriteStream>(worker_queue_, credentials_,
+                                       serializer_bridge_.GetSerializer(),
                                        &grpc_connection_, delegate);
 }
 
+void Datastore::LookupDocuments(
+    const std::vector<DocumentKey> &keys,
+    FSTVoidMaybeDocumentArrayErrorBlock completion) {
+  grpc::ByteBuffer message = serializer_bridge_.ToByteBuffer(
+      serializer_bridge_.CreateLookupRequest(keys));
+
+  auto on_error = [completion](const Status &status) {
+    completion(nil, util::MakeNSError(status));
+  };
+
+  auto on_success =
+      [this, completion](const std::vector<grpc::ByteBuffer> &responses) {
+        Status parse_status;
+        NSArray<FSTMaybeDocument *> *docs =
+            serializer_bridge_.MergeLookupResponses(responses, &parse_status);
+        if (!parse_status.ok()) {
+          completion(nil, util::MakeNSError(parse_status));
+        } else {
+          completion(docs, nil);
+        }
+      };
+
+  WithToken(
+      [this, message, on_success, on_error](const Token &token) {
+        lookup_calls_.push_back(grpc_connection_.CreateStreamingReader(
+            "/google.firestore.v1beta1.Firestore/BatchGetDocuments", token,
+            std::move(message)));
+        GrpcStreamingReader *call = lookup_calls_.back().get();
```

Comment:
Optional reworking of lines 159-162:

```
GRPCStreamingReader *call = grpc_connection_.CreateStreamingReader(...);
lookup_calls_.push_back(call);
```

(Rather than, creating something, adding it to a container, and then looking it up in the container.)